### ISSUE TITLE
Add Expiration field for jwk.Set that is populated by cache-control header

### ISSUE THREAD

**Issue by shaunco** [lestrrat-go/jwx#260]
**Abstract**
It would be great if jwk.Set had an `Expiration` property that is populated by the HTTP `cache-control` header when `FetchHTTP` is used or can be set manually if one of the `Parse` methods is used. Additionally, an 'Origin' property to track where the jwk.Set originated from would be helpful (per the usecase below)

**Describe the proposed solution/change**
Before enumerating available Keys in a Set, one could:
```go
  if set.IsExpired() {
    set, err := set.Refetch()
    if err != nil {
      log.Printf("failed to refetch expired JWKS: %s", err)
      return
    }
  }
```

**Analysis**
Since `FetchHTTP` methods do not make any of the server's headers available to the caller, the only currently alternative is to re-implement `FetchHTTP` outside of jwx.

**Additional context**
Happy to provide a PR if the maintainers agree it would be useful.

**Comment by lestrrat** [comment]
I think in hindsight what I should have done was to remove the FetchHTTP and the like from `jwk`  and put them in a sub package like `github.com/lestrrat-go/jwx/jwk/httputil`. This would have allowed more leeway to add utilities such as the one you described.

It might not be too late to do this. We simply need to change the implementation as follows:

1. We deprecate `"github.com/lestrrat-go/jwx/jwk".FetchHTTP` and the like (but leave the functions as is), and create `"github.com/lestrrat-go/jwx/jwk/httputil".*`
2. In order to break circular referencing we do:
    * Create `github.com/lestrrat-go/jwx/jwk/internal/httputil`, and let `github.com/lestrrat-go/jwx/jwk` and `github.com/lestrrat-go/jwx/jwk/httputil` imports from the internal package.
3. Add whatever else we may need in `github.com/lestrrat-go/jwx/jwk/httputil`

So your code may look like:

```go
import (
  "github.com/lestrrat-go/jwx/jwk"
  "github.com/lestrrat-go/jwx/jwk/httputil"
)
...

set, err := httputil.FetchJWK(....)
...
// XXX1 The underlying implementation defers between JWK key types,
// so probably you need to provide a package-level `IsExpired()` function instead of a method.
//
// XXX2 jwk.IsExpired(set) probably doesn't work, as expiration is a per-key thing?
// It probably needs to be `jwk.IsExpired(key)`
//
// XXX3 While in principle I agree `Refetch(key)` would be nice, this would require that
// each key remember where they originated from -- and you also would need to provide
// APIs to Get/Set source URL on keys who may need to update their source URL.
// Currently I'm not sure how good this will be, but for this example, I'm going to assume
// for the sake of simplicity that each key knows where they come from
//
// XXX4 Finally if keys can be updated, would it be possible that the source
// change the key's underlying type completely? For example, when you first fetched it
// it might have been RSA based, and then on refetch, it could be EC based? I think it's
// quite possible... For the sake of simplicity, we'll assume it's not a problem for this example

if jwk.IsExpired(key) {
  if err := jwk.Refetch(key); err != nil {
      ...
  }
}
```   

So yeah, I think the added convenience is nice, but it seems like there are many subtle
use cases where things need to be fleshed out. 

Perhaps you should come up with a pseudo-code test case before doing a full-blown PR?

**Comment by lestrrat** [comment]
Ah wait, the circular referencing isn't solved with my above comment. Still thinking, but most of my comment still stands.

**Comment by lestrrat** [comment]
Oh, and yes, you mentioned Origin. My bad.

**Comment by shaunco** [comment]
Thank you for the very thoughtful response!

I had implemented this a different way, without the `Origin` and `Refetch()` as Refetch would have required quite a bit of thought around read/write locks. In the end, that seemed better suited for the client to own.

I will submit my PR, but please feel free to decline it if you want to go the route you described above.

**Comment by lestrrat** [comment]
I've been fiddling with this, and am starting to think this is not at the layer of a Key/KeySet. It feels awkward to encode HTTP semantics into the JWK because cache-control and expires are properties of a storage/caching layer, something like the `Cookie Jar` that we have been using for HTTP cookies.

```go
var cache jwk.MemoryCache // in-memory storage
jwk.FetchHTTP(..., WithCache(&cache))
```

The Auth0 library, for example, uses such a.concept to control their JWKs.

**Comment by lestrrat** [comment]
I wonder if this works for you? It uses `github.com/gregjones/httpcache` to implement the HTTP caching.
You would have to *always* use `jwk.Fetch(HTTP)` to get at the JWK to take advantage of the Cache-Control and Expires headers though.

https://github.com/lestrrat-go/jwx/compare/topic/http-cache?expand=1

**Comment by shaunco** [comment]
> I've been fiddling with this, and am starting to think this is not at the layer of a Key/KeySet. It feels awkward to encode HTTP semantics into the JWK because cache-control and expires are properties of a storage/caching layer, something like the `Cookie Jar` that we have been using for HTTP cookies.

I like the MemoryCache model, as it matches a few of the other JWK libraries for go and decouples the expiration from the Set. I assume the TTL would be passed to cache?

**Comment by shaunco** [comment]
> https://github.com/lestrrat-go/jwx/compare/topic/http-cache?expand=1

This looks pretty nice. It would change how I use it so that I end up calling `FetchHTTP` before every `Parse()`/`Validate()`, but given that the call will just hit the cache, it shouldn't add much overhead.

The one catch is that the way `httpcache` package implements thread safety across `Get`, `Set`, and `Delete` looks like any concurrent requests at time of expiration will all end up fetching the JWKS, with each writing the same value back to the map... Likely not a huge deal, depending on cache time and concurrent requests.

**Comment by lestrrat** [comment]
I see, so if your JWK is still valid, it would be overwritten anew on a cache-miss. 
In that case we can probably do

```go
res, _ := httpcl.Do(...)
if v := res.Header.Get(httpcache.XFromCache); v == "" { // v == "" is a cache miss
   set, _ = jwk.Parse(res.Body)
} // otherwise, don't do anything
```

On a side note, I wish JWK by definition only allowed one key instead of one or more keys... Having to deal with a URL possibly pointing to a Key /or/ KeySet is really really really annoying....

Anyways, @shaunco I really would like your use-case as a test. It can contain pseudo code or just be utterly failing as this caching mechanism isn't there yet, but I want to make sure we're on the same wavelength before attempting anymore finetuning.

**Comment by lestrrat** [comment]
Ah, that code snippet was bad, but the idea still stands...

**Comment by shaunco** [comment]
The sync concern was 10 concurrent requests, after the cache has expired, results in:
- `req1` --> `FetchHTTP()` --> Cache miss --> http fetch --> `cache.Set()` --> `Lock()` succeeds --> stores value
- `req2` --> `FetchHTTP()` --> Cache miss --> http fetch --> `cache.Set()` --> `Lock()` blocks until 1 finishes --> stores value
- `req3` --> `FetchHTTP()` --> Cache miss --> http fetch --> `cache.Set()` --> `Lock()` blocks until 1/2 finish --> stores value
- ...

But your point on the JWK still being valid in cache is true as well, although I believe the `httpcl.Do` will check cache before making the request _(didn't verify that)_.

As for my use case, it is a gRPC server that requires authorization via a bearer token. The JWKS URI is set at server startup, but the URI needs refreshed if it expires. Auth0 sets a 15 second expiration on JWKS files (fronted by Cloudflare CDN), which felt excessive, thus my PR having a minimum duration. A few hundred (or more) requests per second come in to the gRPC server, each with an authorization header ("meta data" in gRPC land) containing a bearer token. Those tokens need parsed and validated against the JWKS.

**Comment by lestrrat** [comment]
Right, so your concern for synchronization is about thundering herd. Fair enough. That is usually a concern in the application layer, not on the HTTP layer, so it would have to be dealt atop the HTTP caching.

I'm going to have to ponder for a little while.

Meanwhile, if you could distill your use case that you described as a test, it would make things much easier (obviously, the parts about gRPC isn't required, something barebones using httptest.Server, dummy tokens, dummy jwks, Parse/Validate, and caching)

**Comment by lestrrat** [comment]
re: synchronization, am pretty sure something like this will avoid a thundering herd https://github.com/lestrrat-go/jwx/commit/5eefc72703b0225e3b1390fa1eb5649dd369e641

**Comment by lestrrat** [comment]
I need to jet for the day, but going to leave the following memo for me to remember later:

Possible idea. Use `github.com/gregjones/httpcache` underneath, but add a higher level API like so:

```go
// * Deprecate jwk.FetchXXXX
// * Add storage (or source?) object(s), which knows how to fetch and cache jwk keys.
//        * This may be tricky if we're going to allow both HTTP and filesystem
//.          from the same API
store := NewMemoryStore(
  WithMinimumRefreshInterval(15*time.Minute), // only refresh if the cache object's age is greater than 15 minutes
)
keyset, err := store.Fetch(url) // first time, or when you don't care if the keyset is overwritten

// refresh jwk only if jwk has been updated, check against cache-control/expires
// second argument must be a pointer to key or keyset 
store.Refresh(jwkurl, interface{}, options...)
```

**Comment by lestrrat** [comment]
@shaunco alright, here's my initial take: #262 . 

See [here](https://github.com/lestrrat-go/jwx/pull/262/files#diff-40f65963e86f0e5ec7c15dd282ccb94f2feff9934e698bf2f51c98598ec5b45bR639) for usage.

I still haven't implemented the minimum-time-to-refresh part yet. That's next after I get some feedback.
Let me know what you think

**Comment by shaunco** [comment]
#262 looks good at my first pass. Give me a day to build the usecase as a test.

**Comment by shaunco** [comment]
In case anyone makes it here and is looking for the usecase, it is linked in https://github.com/lestrrat-go/jwx/pull/262#issuecomment-752400686

**Comment by lestrrat** [comment]
fixed by #265
